src/util/DeferredProperty.java

```diff
@@ -0,0 +1,231 @@
+     */
+    public DeferredProperty(T value, Supplier<T> supplier, List<DeferredProperty<?>> observers) {
+        this.value = value;
+        this.supplier = supplier;
+        this.observers = observers;
+        isDirty = value == null;
+    }
+
+    /**
+     * Constructs an instance of a {@link DeferredProperty<T>}.
+     *
+     * @param value The initial value of this property.
+     * @param supplier The non-parametric factory method for updating this property value automatically.
+     */
+    public DeferredProperty(T value, Supplier<T> supplier) {
+        this(value, supplier, new ArrayList<>());
+    }
+
+    /**
+     * Constructs an instance of a {@link DeferredProperty<T>}.
+     *
+     * @param value The initial value of this property.
+     * @param observers The dependent observers to which marking this instance as dirty is propagated to.
+     */
+    public DeferredProperty(T value, List<DeferredProperty<?>> observers) {
+        this(value, null, observers);
+    }
+
+    /**
+     * Constructs an instance of a {@link DeferredProperty<T>}.
+     *
+     * @param value The initial value of this property.
+     * @param observers The dependent observers to which marking this instance as dirty is propagated to.
+     */
+    public DeferredProperty(T value, DeferredProperty<?> ... observers) {
+        this(value, List.of(observers));
+    }
+
+    /**
+     * Constructs an instance of a {@link DeferredProperty<T>}.
+     *
+     * @param value The initial value of this property.
+     */
+    public DeferredProperty(T value) {
+        this(value, new ArrayList<>());
+    }
+
+    /**
+     * Constructs an instance of a {@link DeferredProperty<T>} with an initial value of <code>null</code> and no observers.
+     */
+    public DeferredProperty() {
+        this(null, new ArrayList<>());
+    }
+
+    /**
+     * Bypasses the dirty checks and just returns the value of {@link DeferredProperty#value}.
+     *
+     * @return The current value of {@link DeferredProperty#value}.
+     */
+    public T getValue() {
+        return value;
+    }
+
+    /**
+     * Marks this instance as dirty and only if this instance was initially marked as dirty do we propagate this to its {@link DeferredProperty#observers}.
+     *
+     * @return <code>True</code> if we marked this instance as dirty.
+     */
+    public boolean markAsDirty() {
+        // Only when this instance is clean do we want to make it dirty.
+        if (!isDirty()) {
+            isDirty = true;
+
+            // Mark the observers as dirty as they inherit it from this instance.
+            // As it is impossible to add observers after construction then a livelock would never happen.
+            for (DeferredProperty<?> property : observers) {
+                property.markAsDirty();
+            }
+
+            return true;
+        }
+        return false;
```

Comment:
Since the if-block contains a return statement that is always executed *and* the if-block is more code than the remainder of the function, you might want to revers the check for better readability.

```suggestion
        if (isDirty()) return false;
        
        isDirty = true;

        // Mark the observers as dirty as they inherit it from this instance.
        // As it is impossible to add observers after construction then a livelock would never happen.
        for (DeferredProperty<?> property : observers) {
            property.markAsDirty();
        }

        return true;
```